modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java

```diff
@@ -0,0 +1,6426 @@
+ * <p>There's one more important property of these links: {@code forward(left(i)) == right(i)}. It is called a
+ * <i>triangle invariant</i>. More information on B+Tree structure can easily be found online. Following documentation
+ * concentrates more on specifics of this particular B+Tree implementation.
+ *
+ * <p><h3>General operations.</h3>
+ * This implementation allows for concurrent reads and update. Given that each page locks individually, there are general rules to avoid
+ * deadlocks.
+ * <ul>
+ *     <li>
+ *         Pages within a level always locked from left to right.
+ *     </li>
+ *     <li>
+ *         If there's already a lock on the page of level X then no locks should be acquired on levels less than X.
+ *         In other words, locks are aquired from the bottom to the top. The only exception to this rule is the
+ *         allocation of a new page on a lower level that no one sees yet.
+ *         </li>
+ * </ul>
+ * All basic operations fit into a similar pattern. First, the search is performed ({@link Get}). It goes recursively
+ * from the root to the leaf (if it's needed). On each level several outcomes are possible.
+ * <ul>
+ *     <li>Exact value is found and operation can be completed.</li>
+ *     <li>Insertion point is found and recursive procedure continues on the lower level.</li>
+ *     <li>Insertion point is not found due to concurrent modifications, but retry in the same node is possible.</li>
+ *     <li>Insertion point is not found due to concurrent modifications, but retry in the same node is impossible.</li>
+ * </ul>
+ * All these options, and more, are described in the class {@link Result}. Please refer to its usages for specifics of
+ * each operation. Once the path and the leaf for put/remove is found, the operation is then performed from the bottom
+ * to the top. Specifics are described in corresponding classes ({@link Put}, {@link Remove}).
+ */
+@SuppressWarnings({"ConstantValueVariableUse"})
+public abstract class BplusTree<L, T extends L> extends DataStructure implements IgniteTree<L, T> {
+    /** Empty array. */
+    private static final Object[] EMPTY = {};
+
+    /** Destroy msg. */
+    public static final String CONC_DESTROY_MSG = "Tree is being concurrently destroyed: ";
+
+    /** Number of repetitions to capture a lock in the B+Tree. */
+    // TODO: https://issues.apache.org/jira/browse/IGNITE-16350
+    public static final String IGNITE_BPLUS_TREE_LOCK_RETRIES = "IGNITE_BPLUS_TREE_LOCK_RETRIES";
+
+    /** Number of retries. */
+    private static final int LOCK_RETRIES = getInteger(IGNITE_BPLUS_TREE_LOCK_RETRIES, 1000);
+
+    /** Flag that the tree is destroyed. */
+    private final AtomicBoolean destroyed = new AtomicBoolean(false);
+
+    private final float minFill;
+
+    private final float maxFill;
+
+    /** Meta page id. */
+    protected final long metaPageId;
+
+    private boolean canGetRowFromInner;
+
+    // TODO: https://issues.apache.org/jira/browse/IGNITE-16350
```

Comment:
I'd love to see a comment, not only a link. What's supposed to be done here?

Reply:
Done